File: src/ru/nsu/ccfit/zuev/osu/beatmap/parser/sections/BeatmapGeneralParser.java
package ru.nsu.ccfit.zuev.osu.beatmap.parser.sections;

import ru.nsu.ccfit.zuev.osu.beatmap.BeatmapData;
import ru.nsu.ccfit.zuev.osu.beatmap.constants.BeatmapCountdown;
import ru.nsu.ccfit.zuev.osu.beatmap.constants.SampleBank;

/**
 * A parser for parsing a beatmap's general section.
 */
public class BeatmapGeneralParser extends BeatmapKeyValueSectionParser {
    @Override
    public void parse(BeatmapData data, String line) {
        String[] p = splitProperty(line);

        switch (p[0]) {
            case "AudioFilename" -> data.general.audioFilename = p[1];
            case "AudioLeadIn" -> data.general.audioLeadIn = parseInt(p[1]);
            case "PreviewTime" -> data.general.previewTime = data.getOffsetTime(parseInt(p[1]));
            case "Countdown" -> data.general.countdown = BeatmapCountdown.parse(p[1]);
            case "SampleSet" -> data.general.sampleBank = SampleBank.parse(p[1]);
            case "SampleVolume" -> data.general.sampleVolume = parseInt(p[1]);
            case "StackLeniency" -> data.general.stackLeniency = parseFloat(p[1]);
            case "LetterboxInBreaks" -> data.general.letterboxInBreaks = p[1].equals("1");
            case "Mode" -> data.general.mode = parseInt(p[1]);
        }
    }
}

File: src/com/rian/osu/beatmap/parser/sections/BeatmapGeneralParser.kt
package com.rian.osu.beatmap.parser.sections

import com.rian.osu.beatmap.Beatmap
import com.rian.osu.beatmap.constants.BeatmapCountdown
import com.rian.osu.beatmap.constants.SampleBank
import kotlinx.coroutines.CoroutineScope

/**
 * A parser for parsing a beatmap's general section.
 */
object BeatmapGeneralParser : BeatmapKeyValueSectionParser() {
    override fun parse(beatmap: Beatmap, line: String, scope: CoroutineScope?) = splitProperty(line, scope)?.let {
        when (it.first) {
            "AudioFilename" -> beatmap.general.audioFilename = it.second
            "AudioLeadIn" -> beatmap.general.audioLeadIn = parseInt(it.second)
            "PreviewTime" -> beatmap.general.previewTime = beatmap.getOffsetTime(parseInt(it.second))
            "Countdown" -> beatmap.general.countdown = BeatmapCountdown.parse(it.second)
            "SampleSet" -> beatmap.general.sampleBank = SampleBank.parse(it.second)
            "SampleVolume" -> beatmap.general.sampleVolume = parseInt(it.second)
            "StackLeniency" -> beatmap.general.stackLeniency = parseFloat(it.second)
            "LetterboxInBreaks" -> beatmap.general.letterboxInBreaks = it.second == "1"
            "Mode" -> beatmap.general.mode = parseInt(it.second)
            "SamplesMatchPlaybackRate" -> beatmap.general.samplesMatchPlaybackRate = it.second == "1"
        }
    } ?: throw UnsupportedOperationException("Malformed general property: $line")
}
